core/src/Serial/Kokkos_Serial.cpp

```diff
@@ -59,6 +68,17 @@ void SerialInternal::finalize() {
   }
 
   m_is_initialized = false;
+
+  // guard erasing from all_instances
+  {
+    std::scoped_lock lock(all_instances_mutex);
+    auto it = std::find(all_instances.begin(), all_instances.end(), this);
+    if (it == all_instances.end())
+      Kokkos::abort(
+          "Execution space instance to be removed couldn't be found!");
+    std::swap(*it, all_instances.back());
```

Comment:
Giving it more thought:  you don't even need `swap`, as assignment will do:

```suggestion
    *it = all_instances.back();
```

Sorry about that.  The usual reason to prefer `swap` is when it is non-throwing, but pointer assignment doesn't throw either.